### ISSUE TITLE
N-API Support for hiredis-node

### ISSUE THREAD

**Issue by aruneshchandra** [redis/hiredis-node#134]
The recently announced Node 8 has a new experimental feature called [N-API]( https://nodejs.org/dist/latest-v8.x/docs/api/n-api.html) which is aimed at reducing maintenance cost for node native addons. Checkout this [blog]( https://medium.com/@nodejs/n-api-next-generation-node-js-apis-for-native-modules-169af5235b06) for more details on its benefits. 

`hiredis-node` is one of the top 30 native modules by dependency count, and in order to help the Node.js community make this important transition to N-API, we are hoping you will be able to work with us in order to port `hiredis-node` to support N-API.  Your support and feedback is important in making this effort a success.
 
I am part of the  N-API working group and and would like to talk to you about how you can can get started with N-API and what help we might be able to provide.  I'm available to talk on the phone individually if you'd like.  Alternatively, if you prefer, feel free to jump in our [WG meeting](https://plus.google.com/u/0/events/c0eevtrlajniu7h8cjrdk0f56c8?authkey=COH04YCalJS8Ug) which happens every Thursday at *1.30 Eastern / 10.30 Pacific US time*, to discuss any issues. 
 
Here’s a video of N-API in action 
https://www.youtube.com/watch?v=nmXhJ88nZsk

**Comment by badboy** [comment]
Hey, thanks for reaching out.

If N-API is the way forward, I appreciate that. But at the moment I am not a user of hiredis-node myself nor do I have much time to approach a full rewrite to this new API.
If someone does the effort and provides a full implementation I would be happy to incorporate this here. It won't be me though.

**Comment by aruneshchandra** [comment]
I completely understand your point of view! Will you be willing to talk to us during our  [Weekly WG meeting](https://plus.google.com/u/0/events/c0eevtrlajniu7h8cjrdk0f56c8?authkey=COH04YCalJS8Ug) which happens every **Thursday at 1.30 Eastern / 10.30 Pacific US time**, or one on one over the phone about this ?

**Comment by badboy** [comment]
Not before July, but for that I'll put it on my schedule and let you know a bit in advance.

**Comment by aruneshchandra** [comment]
No problem - we can wait! LMK when we can get some time with you to discuss this a little bit more.

**Comment by aruneshchandra** [comment]
@badboy do you think you have time to talk about this ?

**Comment by badboy** [comment]
Yes! Thanks for reminding me again. I see the next meeting is scheduled for next week, which won't work for me though. How should we get in contact?

**Comment by aruneshchandra** [comment]
When are you available @mhdawson and I can set up a separate call with you on this.

**Comment by badboy** [comment]
@aruneshchandra Sorry for letting this slip. Is there a meeting this week? I would have time to join in.

**Comment by aruneshchandra** [comment]
Yes there is a meeting this week on Thursday at 1.30 Eastern / 10.30 Pacific US time. Here's the hangout link https://plus.google.com/u/0/events/c0eevtrlajniu7h8cjrdk0f56c8?authkey=COH04YCalJS8Ug

Look forward to talking to you :)

**Comment by mhdawson** [comment]
Here is the link to the repo, the readme has instructions on how to use the migration tool:

**Comment by badboy** [comment]
Ok, I took a first stab using the conversion tool and have to say I hoped for a bit more.
Some notes:
* The conversion tool is a mere text replacement, editing the code to refer to non-existing variables here and there. That should be marked more clearly.
* I'm missing a document providing a clearer guide on replacements of the old API. I do understand the conversion tool can't handle everything, but without knowledge about the new APIs it is hard to track down what code to add (the examples are a good start, but don't cover a whole lot)
* The node-addon-api seems to track node.js master and didn't work with my installed node v8.2.1. I had to find that in an issue, it is not documented anywhere.

I don't have any compilable code so far, but will take another stab at it in a couple of days.

**Comment by aruneshchandra** [comment]
//cc: @sampsongao @jasongin

**Comment by NickNaso** [comment]
Hi everyone,
I worked on porting **hiredis** to **N-API**. 
You can find my work here: https://github.com/NickNaso/hiredis-node/tree/napi
If you want experiment with N-API you can create a branch and call it "napi" so I can execute the PR on it.
After that we can iterate to over it and at the end publish a tagged version of**hiredis** like reported here: https://nodejs.org/en/docs/guides/publishing-napi-modules/

**Comment by michael-grunder** [comment]
Hi everyone,

I don't think hiredis-node (or hiredis for that matter) presently have any active maintainers, but I created the [napi](https://github.com/redis/hiredis-node/tree/napi) branch as requested.

I'm not a java/node dev but am happy to help with the process where I can.

Cheers!
Mike

**Comment by berryamin** [comment]
Hi,
  I'll apply for the maintainer role:)
